bench/workgen/workgen.cxx

```diff
@@ -2295,13 +2290,49 @@ Workload::run(WT_CONNECTION *conn)
     return (runner.run(conn));
 }
 
+const std::vector<std::string>
+Workload::get_tables()
+{
+    WorkloadRunner runner(this);
+    std::vector<std::string> uris;
+    {
+        const std::lock_guard<std::mutex> lock(runner._mutex);
+        for (const auto &kv : _context->_internal->_tint) {
+            uris.push_back(kv.first);
+        }
+    }
+    return uris;
+}
+
+void
+Workload::create_table(const std::string &uri)
+{
+    WorkloadRunner runner(this);
+    runner.create_table(uri);
+}
+
 WorkloadRunner::WorkloadRunner(Workload *workload)
     : _workload(workload), _trunners(workload->_threads.size()), _report_out(&std::cout), _start(),
       stopping(false)
 {
     ts_clear(_start);
 }
 
+void
+WorkloadRunner::create_table(const std::string &uri)
+{
+    ContextInternal *icontext = _workload->_context->_internal;
+    {
+        const std::lock_guard<std::mutex> lock(_mutex);
+        if (icontext->_tint.count(uri) != 0)
+            THROW("The table " << uri << " already exists.");
+        tint_t tint = workgen_atomic_add32(&icontext->_tint_last, 1);
+        icontext->_tint[uri] = tint;
+        icontext->_table_names[tint] = uri;
+        icontext->create_all();
```

Comment:
I think there's might be other code that looks at `icontext->_tint` or `icontext->_table_names` or `icontext->_table_runtime` at runtime (that is, during a workload.run).  The one I see is:
```
            recno = workgen_atomic_add64(&_icontext->_table_runtime[tint]._max_recno, 1);
```
`_table_runtime` is altered by `icontext->create_all`, and if that happens during the statement above, it could race (step one: `_icontext->table_runtime` address is gotten in the stmt above, step two: another thread makes a new `table_runtime` and frees the old one (in `ContextInternal::create_all`), step three: first thread accesses `table_runtime`).  So it needs to use the same lock.  I don't see any others, but it's worth taking another look.  And when you add the "select a random table" function, it may need to access one or more of these, those accesses will need to be locked.

Reply:
Thanks, @ddanderson this totally could have happened!
I added some locks here: b1d580904501b4def40efe2f3336609de16abaac. I found two places where we are accessing `table_runtime`.

Then I saw that we are calling `ContextInternal::create_all` at multiple places, I wanted to add a lock inside to protect the internal variables. However, we are already holding a lock when calling it [here](https://github.com/wiredtiger/wiredtiger/pull/8544/commits/80de7cf44a03ddeaf89c7fa97f8386e1467b3cfc#diff-fded55a163215494e7abd497d1b8ee130da02cc03c7f8da23bf54dce10440f31R2331).
In order to have a lock inside `ContextInternal::create_all` and to take into account that we may have a lock defined outside the function, I moved the `mutex` to `recursive_mutex`, let me know what you think, see 80de7cf44a03ddeaf89c7fa97f8386e1467b3cfc.